File: src/lib/fileSystem.ts
import { isTauri } from "./storage";
import type { FileEntry, FileMetadata, UploadFile, AuditLogEntry } from "@/types/fileSystem";

/* ------------------------------------------------------------------ */
/* Sandboxing helper (mirrors the Rust resolve_safe_path)             */
/* ------------------------------------------------------------------ */

export function resolveSafePath(root: string, path: string): string {
  const base = normalize(root);
  const target = normalize(base === "/" ? path : `${base}/${path}`);
  const baseParts = base.split("/").filter(Boolean);
  const targetParts = target.split("/").filter(Boolean);
  if (targetParts.length < baseParts.length) {
    throw new Error("Path traversal detected");
  }
  for (let i = 0; i < baseParts.length; i += 1) {
    if (targetParts[i] !== baseParts[i]) throw new Error("Path traversal detected");
  }
  return "/" + targetParts.join("/");
}

function normalize(p: string): string {
  const parts = p.split("/");
  const out: string[] = [];
  for (const part of parts) {
    if (part === "" || part === ".") continue;
    if (part === "..") {
      if (out.length === 0) throw new Error("Path traversal detected");
      out.pop();
      continue;
    }
    out.push(part);
  }
  return "/" + out.join("/");
}

function parentOf(path: string): string {
  const idx = path.lastIndexOf("/");
  return idx <= 0 ? "/" : path.slice(0, idx);
}

function nameOf(path: string): string {
  return path.slice(path.lastIndexOf("/") + 1);
}

/** Converts a path that may be absolute (starts with projectRoot) into a
 *  path relative to projectRoot, matching the Rust command semantics. */
function toRel(projectRoot: string, path: string): string {
  const r = projectRoot.replace(/\/+$/, "");
  if (path.startsWith(r)) return path.slice(r.length).replace(/^\//, "");
  return path.replace(/^\//, "");
}

/* ------------------------------------------------------------------ */
/* Browser mock filesystem (used when not running inside Tauri)       */
/* ------------------------------------------------------------------ */

interface MockNode {
  is_dir: boolean;
  content: string;
  modified: number;
}

const mockFs = new Map<string, MockNode>();
const auditLogs: AuditLogEntry[] = [];

export function getAuditLogs(): AuditLogEntry[] {
  return auditLogs;
}

function logOp(operation: string, path: string, status: "success" | "error", detail?: string) {
  const entry: AuditLogEntry = {
    timestamp: new Date().toISOString(),
    operation,
    path,
    status,
    detail,
  };
  auditLogs.push(entry);
  // eslint-disable-next-line no-console
  console.log(`[audit] ${entry.operation} ${entry.status} ${entry.path}`, detail ?? "");
}

function seed(root: string) {
  if (mockFs.has(root)) return;
  const now = Date.now();
  const put = (path: string, is_dir: boolean, content = "") =>
    mockFs.set(path, { is_dir, content, modified: now });
  put(root, true);
  put(`${root}/README.md`, false, "# OpenComputer\n\nWelcome to your project.\n");
  put(
    `${root}/package.json`,
    false,
    JSON.stringify({ name: "demo-project", version: "1.0.0" }, null, 2),
  );
  put(`${root}/src`, true);
  put(
    `${root}/src/App.tsx`,
    false,
    "export default function App() {\n  return <h1>Hello</h1>;\n}\n",
  );
  put(`${root}/src/main.tsx`, false, "import App from './App';\n");
  put(`${root}/docs`, true);
  put(`${root}/docs/notes.md`, false, "## Notes\n\nPhase 2 file system.\n");
}

function mockChildren(dir: string): FileEntry[] {
  const entries: FileEntry[] = [];
  for (const [path, node] of mockFs) {
    if (parentOf(path) === dir) {
      entries.push({
        name: nameOf(path),
        path,
        is_dir: node.is_dir,
        size: node.is_dir ? 0 : node.content.length,
        modified: node.modified,
      });
    }
  }
  entries.sort((a, b) => {
    if (a.is_dir !== b.is_dir) return a.is_dir ? -1 : 1;
    return a.name.localeCompare(b.name);
  });
  return entries;
}

function mockCollectDescendants(dir: string, out: string[]) {
  for (const [path, node] of mockFs) {
    if (path !== dir && path.startsWith(dir + "/") && node.is_dir) {
      mockCollectDescendants(path, out);
    }
    if (path.startsWith(dir + "/")) out.push(path);
  }
}

/* ------------------------------------------------------------------ */
/* Tauri invoke wrapper                                               */
/* ------------------------------------------------------------------ */

async function tauriInvoke<T>(cmd: string, args: Record<string, unknown>): Promise<T> {
  const { invoke } = await import("@tauri-apps/api/core");
  return invoke<T>(cmd, args);
}

/* ------------------------------------------------------------------ */
/* Public API                                                         */
/* ------------------------------------------------------------------ */

export async function listDirectory(path: string, projectRoot: string): Promise<FileEntry[]> {
  if (isTauri()) {
    try {
      const res = await tauriInvoke<FileEntry[]>("list_directory", { path, projectRoot });
      logOp("list_directory", path, "success");
      return res;
    } catch (e) {
      logOp("list_directory", path, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const safe = resolveSafePath(projectRoot, toRel(projectRoot, path));
    const entries = mockChildren(safe);
    logOp("list_directory", safe, "success");
    return entries;
  } catch (e) {
    logOp("list_directory", path, "error", String(e));
    throw e;
  }
}

export async function readFile(path: string, projectRoot: string): Promise<string> {
  if (isTauri()) {
    try {
      const res = await tauriInvoke<string>("read_file", { path, projectRoot });
      logOp("read_file", path, "success");
      return res;
    } catch (e) {
      logOp("read_file", path, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const safe = resolveSafePath(projectRoot, toRel(projectRoot, path));
    const node = mockFs.get(safe);
    if (!node || node.is_dir) throw new Error("File not found");
    logOp("read_file", safe, "success");
    return node.content;
  } catch (e) {
    logOp("read_file", path, "error", String(e));
    throw e;
  }
}

export async function writeFile(
  path: string,
  content: string,
  projectRoot: string,
): Promise<void> {
  if (isTauri()) {
    try {
      await tauriInvoke<void>("write_file", { path, content, projectRoot });
      logOp("write_file", path, "success");
      return;
    } catch (e) {
      logOp("write_file", path, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const safe = resolveSafePath(projectRoot, toRel(projectRoot, path));
    const parent = parentOf(safe);
    if (!mockFs.has(parent)) throw new Error(`Parent directory does not exist: ${parent}`);
    const existing = mockFs.get(safe);
    mockFs.set(safe, {
      is_dir: false,
      content,
      modified: Date.now(),
    });
    if (!existing) logOp("create_file", safe, "success");
    else logOp("write_file", safe, "success");
  } catch (e) {
    logOp("write_file", path, "error", String(e));
    throw e;
  }
}

export async function createFolder(path: string, projectRoot: string): Promise<void> {
  if (isTauri()) {
    try {
      await tauriInvoke<void>("create_folder", { path, projectRoot });
      logOp("create_folder", path, "success");
      return;
    } catch (e) {
      logOp("create_folder", path, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const safe = resolveSafePath(projectRoot, toRel(projectRoot, path));
    const parent = parentOf(safe);
    if (!mockFs.has(parent)) throw new Error(`Parent directory does not exist: ${parent}`);
    if (mockFs.has(safe)) throw new Error("Path already exists");
    mockFs.set(safe, { is_dir: true, content: "", modified: Date.now() });
    logOp("create_folder", safe, "success");
  } catch (e) {
    logOp("create_folder", path, "error", String(e));
    throw e;
  }
}

export async function renameFile(
  oldPath: string,
  newPath: string,
  projectRoot: string,
): Promise<void> {
  return moveFile(oldPath, newPath, projectRoot);
}

export async function moveFile(
  source: string,
  destination: string,
  projectRoot: string,
): Promise<void> {
  if (isTauri()) {
    try {
      await tauriInvoke<void>("move_file", { source, destination, projectRoot });
      logOp("move_file", `${source} -> ${destination}`, "success");
      return;
    } catch (e) {
      logOp("move_file", `${source} -> ${destination}`, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const src = resolveSafePath(projectRoot, toRel(projectRoot, source));
    const dst = resolveSafePath(projectRoot, toRel(projectRoot, destination));
    if (!mockFs.has(src)) throw new Error("Source does not exist");
    const toMove = [src];
    mockCollectDescendants(src, toMove);
    const moved: Array<[string, MockNode]> = [];
    for (const p of toMove) {
      const node = mockFs.get(p)!;
      const rel = p.slice(src.length);
      moved.push([dst + rel, node]);
      mockFs.delete(p);
    }
    for (const [np, node] of moved) mockFs.set(np, node);
    logOp("move_file", `${src} -> ${dst}`, "success");
  } catch (e) {
    logOp("move_file", `${source} -> ${destination}`, "error", String(e));
    throw e;
  }
}

export async function deleteFile(path: string, projectRoot: string): Promise<void> {
  if (isTauri()) {
    try {
      await tauriInvoke<void>("delete_file", { path, projectRoot });
      logOp("delete_file", path, "success");
      return;
    } catch (e) {
      logOp("delete_file", path, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const safe = resolveSafePath(projectRoot, toRel(projectRoot, path));
    if (!mockFs.has(safe)) throw new Error("Path does not exist");
    const toDelete = [safe];
    mockCollectDescendants(safe, toDelete);
    for (const p of toDelete) mockFs.delete(p);
    logOp("delete_file", safe, "success");
  } catch (e) {
    logOp("delete_file", path, "error", String(e));
    throw e;
  }
}

export async function uploadFiles(
  files: UploadFile[],
  targetDir: string,
  projectRoot: string,
): Promise<void> {
  if (isTauri()) {
    try {
      await tauriInvoke<void>("upload_files", { files, targetDir, projectRoot });
      logOp("upload_files", targetDir, "success", `${files.length} file(s)`);
      return;
    } catch (e) {
      logOp("upload_files", targetDir, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const safeDir = resolveSafePath(projectRoot, toRel(projectRoot, targetDir));
    if (!mockFs.has(safeDir) || !mockFs.get(safeDir)!.is_dir) {
      throw new Error("Target directory does not exist");
    }
    for (const f of files) {
      const dest = `${safeDir}/${f.name}`;
      mockFs.set(dest, { is_dir: false, content: f.content, modified: Date.now() });
    }
    logOp("upload_files", safeDir, "success", `${files.length} file(s)`);
  } catch (e) {
    logOp("upload_files", targetDir, "error", String(e));
    throw e;
  }
}

export async function getFileMetadata(
  path: string,
  projectRoot: string,
): Promise<FileMetadata> {
  if (isTauri()) {
    try {
      const res = await tauriInvoke<FileMetadata>("get_file_metadata", { path, projectRoot });
      logOp("get_file_metadata", path, "success");
      return res;
    } catch (e) {
      logOp("get_file_metadata", path, "error", String(e));
      throw e;
    }
  }
  try {
    seed(projectRoot);
    const safe = resolveSafePath(projectRoot, toRel(projectRoot, path));
    const node = mockFs.get(safe);
    if (!node) throw new Error("Path does not exist");
    const meta: FileMetadata = {
      name: nameOf(safe),
      path: safe,
      is_dir: node.is_dir,
      size: node.is_dir ? 0 : node.content.length,
      modified: node.modified,
    };
    logOp("get_file_metadata", safe, "success");
    return meta;
  } catch (e) {
    logOp("get_file_metadata", path, "error", String(e));
    throw e;
  }
}

File: src-tauri/src/file_system.rs
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use tauri::AppHandle;

#[derive(Serialize, Deserialize, Clone)]
pub struct FileEntry {
    pub name: String,
    pub path: String,
    pub is_dir: bool,
    pub size: u64,
    pub modified: i64,
}

#[derive(Serialize, Deserialize)]
pub struct FileMetadata {
    pub name: String,
    pub path: String,
    pub is_dir: bool,
    pub size: u64,
    pub modified: i64,
}

#[derive(Serialize, Deserialize)]
pub struct UploadFile {
    pub name: String,
    pub content: String,
}

/// Resolves `path` against `project_root` and ensures it stays inside the root.
/// Rejects any path containing `..` or that escapes the root directory.
fn resolve_safe_path(path: &str, project_root: &str) -> Result<PathBuf, String> {
    if path.contains("..") {
        return Err("Path traversal detected".to_string());
    }
    let root = PathBuf::from(project_root)
        .canonicalize()
        .map_err(|e| format!("Invalid project root: {e}"))?;

    let candidate = if Path::new(path).is_absolute() {
        PathBuf::from(path)
    } else {
        root.join(path)
    };

    let canonical = candidate.canonicalize().unwrap_or_else(|_| candidate.clone());
    if !canonical.starts_with(&root) {
        return Err("Path traversal detected".to_string());
    }
    Ok(candidate)
}

fn modified_ms(meta: &fs::Metadata) -> i64 {
    meta.modified()
        .ok()
        .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

/// Appends an operation to the Security Audit Log (Phase 6 consumer).
fn log_operation(app: &AppHandle, operation: &str, path: &str, status: &str) {
    if let Ok(dir) = app.path().app_config_dir() {
        let _ = fs::create_dir_all(&dir);
        let log_path = dir.join("opencomputer-audit.log");
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_secs())
            .unwrap_or(0);
        let line = format!("{} {} {} {}\n", now, operation, status, path);
        if let Ok(mut f) = fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(log_path)
        {
            let _ = f.write_all(line.as_bytes());
        }
    }
}

#[tauri::command]
pub fn list_directory(
    app: AppHandle,
    path: String,
    project_root: String,
) -> Result<Vec<FileEntry>, String> {
    let full = resolve_safe_path(&path, &project_root)?;
    let read = fs::read_dir(&full).map_err(|e| format!("Cannot read directory: {e}"))?;
    let mut entries: Vec<FileEntry> = Vec::new();
    for entry in read.flatten() {
        let meta = match entry.metadata() {
            Ok(m) => m,
            Err(_) => continue,
        };
        let is_dir = meta.is_dir();
        entries.push(FileEntry {
            name: entry.file_name().to_string_lossy().to_string(),
            path: entry.path().to_string_lossy().to_string(),
            is_dir,
            size: if is_dir { 0 } else { meta.len() },
            modified: modified_ms(&meta),
        });
    }
    entries.sort_by(|a, b| {
        if a.is_dir != b.is_dir {
            return a.is_dir.cmp(&b.is_dir).reverse();
        }
        a.name.to_lowercase().cmp(&b.name.to_lowercase())
    });
    log_operation(&app, "list_directory", &path, "success");
    Ok(entries)
}

#[tauri::command]
pub fn read_file(
    app: AppHandle,
    path: String,
    project_root: String,
) -> Result<String, String> {
    let full = resolve_safe_path(&path, &project_root)?;
    let content = fs::read_to_string(&full).map_err(|e| format!("Cannot read file: {e}"))?;
    log_operation(&app, "read_file", &path, "success");
    Ok(content)
}

#[tauri::command]
pub fn write_file(
    app: AppHandle,
    path: String,
    content: String,
    project_root: String,
) -> Result<(), String> {
    let full = resolve_safe_path(&path, &project_root)?;
    if let Some(parent) = full.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("Cannot create parent: {e}"))?;
    }
    fs::write(&full, content).map_err(|e| format!("Cannot write file: {e}"))?;
    log_operation(&app, "write_file", &path, "success");
    Ok(())
}

#[tauri::command]
pub fn create_folder(
    app: AppHandle,
    path: String,
    project_root: String,
) -> Result<(), String> {
    let full = resolve_safe_path(&path, &project_root)?;
    fs::create_dir_all(&full).map_err(|e| format!("Cannot create folder: {e}"))?;
    log_operation(&app, "create_folder", &path, "success");
    Ok(())
}

#[tauri::command]
pub fn rename_file(
    app: AppHandle,
    old_path: String,
    new_path: String,
    project_root: String,
) -> Result<(), String> {
    let src = resolve_safe_path(&old_path, &project_root)?;
    let dst = resolve_safe_path(&new_path, &project_root)?;
    fs::rename(&src, &dst).map_err(|e| format!("Cannot rename: {e}"))?;
    log_operation(&app, "rename_file", &old_path, "success");
    Ok(())
}

#[tauri::command]
pub fn delete_file(
    app: AppHandle,
    path: String,
    project_root: String,
) -> Result<(), String> {
    let full = resolve_safe_path(&path, &project_root)?;
    let result = if full.is_dir() {
        fs::remove_dir_all(&full)
    } else {
        fs::remove_file(&full)
    };
    result.map_err(|e| format!("Cannot delete: {e}"))?;
    log_operation(&app, "delete_file", &path, "success");
    Ok(())
}

#[tauri::command]
pub fn move_file(
    app: AppHandle,
    source: String,
    destination: String,
    project_root: String,
) -> Result<(), String> {
    let src = resolve_safe_path(&source, &project_root)?;
    let dst = resolve_safe_path(&destination, &project_root)?;
    if let Some(parent) = dst.parent() {
        fs::create_dir_all(parent).map_err(|e| format!("Cannot create parent: {e}"))?;
    }
    fs::rename(&src, &dst).map_err(|e| format!("Cannot move: {e}"))?;
    log_operation(&app, "move_file", &source, "success");
    Ok(())
}

#[tauri::command]
pub fn upload_files(
    app: AppHandle,
    files: Vec<UploadFile>,
    target_dir: String,
    project_root: String,
) -> Result<(), String> {
    let dir = resolve_safe_path(&target_dir, &project_root)?;
    fs::create_dir_all(&dir).map_err(|e| format!("Cannot create target: {e}"))?;
    for f in files {
        let dest = dir.join(&f.name);
        if !dest.starts_with(&dir) {
            return Err("Path traversal detected".to_string());
        }
        fs::write(&dest, f.content).map_err(|e| format!("Cannot write upload: {e}"))?;
    }
    log_operation(&app, "upload_files", &target_dir, "success");
    Ok(())
}

#[tauri::command]
pub fn get_file_metadata(
    app: AppHandle,
    path: String,
    project_root: String,
) -> Result<FileMetadata, String> {
    let full = resolve_safe_path(&path, &project_root)?;
    let meta = fs::metadata(&full).map_err(|e| format!("Cannot stat: {e}"))?;
    let is_dir = meta.is_dir();
    let result = FileMetadata {
        name: full
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default(),
        path: full.to_string_lossy().to_string(),
        is_dir,
        size: if is_dir { 0 } else { meta.len() },
        modified: modified_ms(&meta),
    };
    log_operation(&app, "get_file_metadata", &path, "success");
    Ok(result)
}
